c10/core/impl/COW.cpp

```diff
@@ -76,38 +108,95 @@ c10::intrusive_ptr<StorageImpl> lazy_clone_storage(StorageImpl& storage) {
   //
   //    No locking is required in this case.
 
-  std::optional<DataPtr> new_data_ptr; // must be set below
+  std::optional<DataPtr> new_data_ptr_opt; // must be set below
 
   if (has_simple_data_ptr(storage)) {
     // Case 1) We have a simple data pointer: wrap it.
     std::unique_ptr<void, DeleterFnPtr> original_ctx =
         storage._mutable_data_ptr_no_checks().move_context();
 
     // Save this for the result.
-    new_data_ptr = make_data_ptr(
-        data_ptr, *new cow::COWDeleterContext(std::move(original_ctx)));
+    new_data_ptr_opt = make_data_ptr(
+        data_ptr,
+        *new cow::COWDeleterContext(
+            std::move(original_ctx), storage.device(), storage.allocator()));
 
     // Update this storage to the new copy on write context.
-    storage.set_data_ptr_noswap(copy_data_ptr(*new_data_ptr));
+    storage.set_data_ptr_noswap(copy_data_ptr(*new_data_ptr_opt));
   } else if (is_cow_data_ptr(data_ptr)) {
     // Case 2): there is already a copy on write context. Just return a
     // new storage impl.
-    new_data_ptr = copy_data_ptr(data_ptr);
+    new_data_ptr_opt = copy_data_ptr(data_ptr);
   } else {
     // Case 3) There is a context and it's not copy-on-write. Nothing
     // we can do here.
     return nullptr;
   }
 
-  TORCH_INTERNAL_ASSERT(new_data_ptr.has_value());
+  TORCH_INTERNAL_ASSERT(new_data_ptr_opt.has_value());
+
+  c10::Allocator* allocator = storage.allocator();
+  c10::DeviceType device_type = storage.device_type();
+
+  if (device_opt.has_value()) {
+    allocator = allocator_opt.value();
```

Comment:
If `allocator_opt` does not contain a value then this will throw. Should default to either the storage's allocator or the default allocator for the device?